### PR TITLE
[CGFCOLLECTOR] fixes and more tests

### PR DESCRIPTION
**Pull request by gomfol12**
Hi,
I ran the `cgfcollector` against a bigger code base (icon) and noticed some issues. This PR fixes them and adds tests.

- fix a crash when `null()` is used to initialize a pointer outside a procedure definition.
- fix a crash when the argument position of a procedure was not correctly determined.
- fix edges not found for calls to custom operator with irregular capitalization.
- fix wrapper scripts (`cgfcollector_comp_wrapper.sh` and cgfcollector_link_wrapper.sh) not honoring `--skip-gen-bin` option.

- add test case for `null()` usage outside procedure definition.
- add test case for custom operators with irregular capitalization.
- add test script for `cgfcollector_comp_wrapper.sh`
- add test script for `cgfcollector_link_wrapper.sh`
- updated top comment for `cgfcollector_comp_wrapper.sh` and `cgfcollector_link_wrapper.sh` to document all available options.

Let me know if I should split this PR. The fixes itself are small, so I don't know if this justifies the extra effort.

### CODE REVIEW

File: cgfcollector/src/FortranUtil.cpp [hunk 1]
```diff
@@ -171,31 +171,35 @@ std::string getOperatorStringFromExpr(const Expr* expr) {
...
+                          [](const Expr::EQV&) { return std::string("EQV"); },
+                          [](const Expr::NEQV&) { return std::string("NEQV"); },
+
+                          [](const Expr::DefinedUnary& op) { return std::get<DefinedOpName>(op.t).v.ToString(); },
+                          [](const Expr::DefinedBinary& op) { return std::get<DefinedOpName>(op.t).v.ToString(); },
+
+                          [](const auto&) -> std::string { return std::string("UNKNOWN_OPERATOR"); }},
+                 expr->u);
+
+  std::transform(out.begin(), out.end(), out.begin(), ::toupper);
```
Comment (TimHeldmann) [inline cgfcollector/src/FortranUtil.cpp]: I always disliked C++ String handling...
Is this to uppercase potential returns of Expr::DefinedUnary and Expr::DefinedBinary?
Because if so, I would prefer to have the uppercasing only done in these cases.

File: cgfcollector/src/ParseTreeVisitor.cpp [hunk 2]
```diff
@@ -18,6 +18,9 @@ namespace metacg::cgfcollector {
 
 template <typename Iterable, typename Extractor>
 void ParseTreeVisitor::handleDummyArgs(const Iterable& items, Extractor extract) {
+  if (currentProcedures.empty())
```
Comment (TimHeldmann) [inline cgfcollector/src/ParseTreeVisitor.cpp]: This condition repeats throughout the code.
Can you abort traversal of the currentProcedure if it is empty, to not need to deal with this in every callback?

File: cgfcollector/tools/cgfcollector_comp_wrapper.sh.in [hunk 3]
```diff
@@ -14,14 +14,15 @@
 # Usage:
 #     cgfcollector_comp_wrapper.sh [flang options, source files, plugin options]
 #
-# Plugin-specific options:
+# Options:
```
Comment (TimHeldmann) [inline cgfcollector/tools/cgfcollector_comp_wrapper.sh.in:27]: Which part of the PR do the changes for this file belong to?

File: cgfcollector/tools/cgfcollector_link_wrapper.sh.in [hunk 4]
```diff
@@ -4,6 +4,9 @@
 #
 # Usage:
 #     cgfcollector_link_wrapper.sh [flang options and object files]
+#
```
Comment (TimHeldmann) [inline cgfcollector/tools/cgfcollector_link_wrapper.sh.in:7]: Similar as above, how does this pertain to the --skip-gen-bin fix?

**Review by TimHeldmann — CHANGES_REQUESTED** [review]

File: cgfcollector/src/FortranUtil.cpp (hunk 1 above)
Comment (gomfol12) [reply cgfcollector/src/FortranUtil.cpp]: done

File: cgfcollector/tools/cgfcollector_comp_wrapper.sh.in (hunk 3 above)
Comment (gomfol12) [reply cgfcollector/tools/cgfcollector_comp_wrapper.sh.in:27]: To none. I have put it in a separate commit and updated the description of this PR to also include this documentation change.

File: cgfcollector/tools/cgfcollector_link_wrapper.sh.in (hunk 4 above)
Comment (gomfol12) [reply cgfcollector/tools/cgfcollector_link_wrapper.sh.in:7]: same as above

File: cgfcollector/tools/cgfcollector_comp_wrapper.sh.in [hunk 5]
```diff
@@ -14,14 +14,15 @@
...
 #     -d, --dot             # Additionally generate a DOT file. This is mostly used for debugging.
 #     --skip-gen-bin        # Skip the final compilation step. Only generate the call graph.
 #     -v, --verbose         # Print additional information during the generation process.
 #     --graph-name          # Set internal graph name
 #     --include-intrinsics  # Include intrinsic calls in the generated call graph.
 
 flang_bin=${CGFCOLLECTOR_FLANG_BIN:-"flang-new"}
```
Comment (pearzt) [inline cgfcollector/tools/cgfcollector_comp_wrapper.sh.in]: I think this slipped by us in the review of the first PR. In recent versions of LLVM, the flang binary is just called `flang`. As it seems likely that this won't change in the future, maybe we should use the new name as the default here (and also in the README)?

Something related that I struggle to understand: This script is designed to be used with `CMAKE_Fortran_COMPILER_LAUNCHER`, right? Why does it need to know the name of the compiler then? Shouldn't that just be in `argv[1]`?

File: cgfcollector/src/ParseTreeVisitor.cpp (hunk 2 above)
Comment (gomfol12) [reply cgfcollector/src/ParseTreeVisitor.cpp]: I look into it and I realized that I was overly cautious with this check. So I removed the unnecessary once. 
This includes the check in ` handleDummyArgs`, `handleEndFuncSubStmt`, `Post(const MainProgram&)`, `Post(const SubroutineSubprogram&)`, `Post(const EndFunctionStmt&)`, `Post(const EndSubroutineStmt&)`, `Pre(const StmtFunctionStmt& s)` and `Post(const StmtFunctionStmt& s)`.
 In these cases I always insert into `currentProcedures` before accessing it.

The remaining once I need to keep because the callbacks refer to Fortran constructs that can be used in and outside a procedure.

About the second part. Not all the checks are on the same path thought the tree, so aborting traversal would not save any checks. Also for some cases I want to further traverse if I am not in a procedure (e.g. to collect type information).

File: cgfcollector/tools/cgfcollector_comp_wrapper.sh.in (hunk 5 above)
Comment (gomfol12) [reply cgfcollector/tools/cgfcollector_comp_wrapper.sh.in]: First part. Done. I now use primarily flang.

Second part. Originally this script was designed to be used as a drop-in replacement for a Fortran compiler. I then later realized that for CMake projects the `CMAKE_Fortran_COMPILER_LAUNCHER` option would make the usage easier. So this script can be used as both a compiler wrapper and a drop-in compiler replacement.

I would like to keep both functionalities, because for autotools projects I just can set the `FC` variable to my compiler wrapper and it works.

Comment (pearzt) [reply cgfcollector/tools/cgfcollector_comp_wrapper.sh.in]: Alright, makes sense. Please document these two functionalities of the script in the comments and the README, though. Also, instead of relying on the undocumented `CGFCOLLECTOR_FLANG_BIN`, we could just check which binary is installed:

```cmake
find_program(FLANG_EXECUTABLE
    NAMES flang flang-new
    HINTS ${LLVM_TOOLS_BINARY_DIR}
    NO_DEFAULT_PATH
)
```

**Review by TimHeldmann — APPROVED** [review]

File: cgfcollector/tools/cgfcollector_comp_wrapper.sh.in (hunk 5 above)
Comment (gomfol12) [reply cgfcollector/tools/cgfcollector_comp_wrapper.sh.in]: I documented these two functionalities in the script and README.
I would like to keep `CGFCOLLECTOR_FLANG_BIN` because it is useful if you want to run other Flang wrappers (like `mpif90`). So, I also featured it more in the documentation.

Comment (pearzt) [reply cgfcollector/tools/cgfcollector_comp_wrapper.sh.in]: Ah, right, that makes sense, I didn't think of `mpif90`.

File: cgfcollector/tools/cgfcollector_comp_wrapper.sh.in [hunk 6]
```diff
@@ -1,18 +1,28 @@
...
-# keeping only those compatible with `-fc1`, and invokes flang with
-# the plugin to generate call graphs. Finally, it invokes flang again
+# keeping only those compatible with `-fc1`, and invokes Flang with
+# the plugin to generate call graphs. Finally, it invokes Flang again
 # with the original arguments to perform the actual compilation.
 #
 # Usage:
-#     cgfcollector_comp_wrapper.sh [flang options, source files, plugin options]
+#     1. Compiler wrapper:
+#        Pas the compiler as the first argument. This takes precedence.
```
Comment (pearzt) [inline cgfcollector/tools/cgfcollector_comp_wrapper.sh.in]: typo: Pas -> Pass

**Review by pearzt — CHANGES_REQUESTED** [review]

**Review by pearzt — APPROVED** [review]
